### PR TITLE
Validate Cloud SDK and Java components before use

### PR DESCRIPTION
**Pull request by meltsufin**
@etanshaul @patflynn @loosebazooka @elharo 

I split the validation up into validating Cloud SDK separately from the Java components, and created a separate exception for when Cloud SDK is not found.
I also made the checks happen by default in the methods where the corresponding components are used.

### CODE REVIEW

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java [hunk 1]
```diff
@@ -131,10 +134,19 @@ public void runAppCommand(List<String> args) throws ProcessRunnerException {
...
-   * @throws InvalidPathException when Python can't be located
-   * @throws ProcessRunnerException when process runner encounters an error
+   * @throws InvalidPathException      when Python can't be located
+   * @throws ProcessRunnerException    when process runner encounters an error
+   * @throws CloudSdkNotFoundException when the Cloud SDK is not installed where expected
+   * @throws AppEngineException        when dev_appserver.py cannot be found
    */
   public void runDevAppServerCommand(List<String> args) throws ProcessRunnerException {
+    validateCloudSdk();
+    if (!Files.isRegularFile(getDevAppServerPath())) {
```
Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: perhaps this could be in validateDevAppServer()?

Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: maybe wrap this in a validateDevAppServer() (or something like that) to keep the validation consistent?

Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Is dev_appserver.py always installed with Cloud SDK? If so, I can roll this into validateCloudSdk().

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java [hunk 2]
```diff
@@ -231,36 +248,39 @@ public Path getJarPath(String jarName) {
...
+   * Checks whether the Cloud SDK Path with is valid.
    *
-   * @throws AppEngineJavaComponentsNotInstalledException when the App Engine Java 
-   *     components are not installed in the Cloud SDK
-   * @throws AppEngineException when the Cloud SDK is not installed where expected
+   * @throws CloudSdkNotFoundException when the Cloud SDK is not installed where expected
    */
-  public void validate() throws AppEngineException {
+  public void validateCloudSdk() throws CloudSdkNotFoundException {
     if (sdkPath == null) {
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: IT feels like each one of these conditions is different and a client would probably want to handle them differently.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 1 above)
Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: also are we sure that dev app server doesn't require the java components?

Comment (loosebazooka) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: it's part of the appengine components

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package com.google.cloud.tools.appengine.cloudsdk;
+
+import com.google.cloud.tools.appengine.api.AppEngineException;
+
+/**
+ * User needs to use a valid Cloud SDK installation.
+ */
+public class CloudSdkNotFoundException extends AppEngineException {
```
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: I'm not seeing a lot of value in the AppEngineException class of exceptions anymore.  It seems to be used by all exceptions that could possibly occur in the library rather than the unexpected errors caused by library deps.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java [hunk 4]
```diff
@@ -131,10 +134,19 @@ public void runAppCommand(List<String> args) throws ProcessRunnerException {
...
-   * @throws ProcessRunnerException when process runner encounters an error
+   * @throws InvalidPathException      when Python can't be located
+   * @throws ProcessRunnerException    when process runner encounters an error
+   * @throws CloudSdkNotFoundException when the Cloud SDK is not installed where expected
+   * @throws AppEngineException        when dev_appserver.py cannot be found
    */
   public void runDevAppServerCommand(List<String> args) throws ProcessRunnerException {
+    validateCloudSdk();
+    if (!Files.isRegularFile(getDevAppServerPath())) {
+      throw new AppEngineException(
```
Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Should this have a specific exception subclass like the others?

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 1 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Ss the app-engine-java component necessary and sufficient for running Java apps with dev_appserver.py?

Comment (patflynn) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: the dev-appserver is always included with the cloud sdk, the
app-engine-java component is not required for running the dev-appserver,
but apparently you need to invoke it correctly so that it'll auto-install
the java component when it detects a java runtime is needed.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 2 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: I think for now we just want to let clients know if the Cloud SDK is valid or not. Details are in the exception message.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java (hunk 3 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: It could be useful for a client that just needs to know broadly if there is a problem coming from the lib.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 1 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: We should just do the auto-install flag then.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java [hunk 5]
```diff
@@ -156,8 +168,13 @@ public void runDevAppServerCommand(List<String> args) throws ProcessRunnerExcept
 
   /**
    * Executes an App Engine SDK CLI command.
+   *
+   * @throws AppEngineJavaComponentsNotInstalledException when the App Engine Java components are
+   *                                                      not installed in the Cloud SDK
    */
   public void runAppCfgCommand(List<String> args) throws ProcessRunnerException {
```
Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: shouldn't this method also `validateCloudSdk()` ?

Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Technically, the check for the existence of the jar is sufficient.

Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Yeah, that's true.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java (hunk 3 above)
Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: Might be a good time to change to a checked exception then since these are all environmental problems, not programming errors.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 2 above)
Comment (patflynn) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: That's not useful for the IDE use-cases. We want to present actionable
errors to the user and we don't want to pass through your message strings
since we support localization.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 1 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Found out that auto-install currently doesn't work for Java runtime.

**Comment by etanshaul** [comment]
lgtm

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java (hunk 3 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: I still would rather do it in a separate PR, since it's would be more disruptive to clients.

Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: I could wait on making the old AppEngineException checked, but this one should be checked form the get go since the point of this PR is that clients should be catching and handling this exception. Otherwise it's really not useful.

The alternative would be to rethink throwing the exception and have it return a boolean instead.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+  CloudSdkNotFoundException(String message) {
```
Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: Please make this public as well per our offline discussion (so we can mock these exceptions in client unit tests).

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java (hunk 3 above)
Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdkNotFoundException.java]: I agree that making the exception checked is the right approach here. I just want to make that change separate because it affects a lot of places in the code.

**Comment by meltsufin** [comment]
@etanshaul can you take a look one more time before I merge?

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java [hunk 7]
```diff
@@ -279,7 +279,8 @@ public void validateAppEngineJavaComponents()
       throws AppEngineJavaComponentsNotInstalledException {
     if (!Files.isDirectory(getJavaAppEngineSdkPath())) {
       throw new AppEngineJavaComponentsNotInstalledException(
-          "Validation Error: Java App Engine components not installed");
+          "Validation Error: Java App Engine components not installed."
+              + " Fix by running 'gcloud components install app-engine-java' on command-line.");
```
Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: Didn't we say we wouldn't suggest fixes here in the exception message, but instead let the clients determine how to handle it? One reason would be if clients auto download the component, it might not make sense to have this here.

**Comment by etanshaul** [comment]
Other than my comment on the updated exception message, lgtm.

File: src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java (hunk 7 above)
Comment (etanshaul) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: I am ok with it though - I don't see it as a huge issue. So feel free to merge once you decide what you want to do it it. Also please release the lib afterwards :)

Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: I think when the auto-download is enabled, this exception simply won't show up. My motivation for including the fix in the message is to just short-circuit the resolution of the problem for users, if the library clients somehow fail to handle the exception correctly or decide to simply display the message contained in the exception.

Comment (elharo) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: I'm with @etanshaul on this one. This library should be quite agnostic about whether there's a user or not.

Comment (meltsufin) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: I would revisit this when these exceptions are made checked. For the time being, however, since the clients are not forced to handle the exception, I'd rather have this in there to potentially save time for users encountering this issue.

Comment (patflynn) [inline src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java]: LGTM
